src/coreclr/System.Private.CoreLib/src/Microsoft/Win32/OAVariantLib.cs

```diff
@@ -26,102 +30,164 @@ internal static unsafe partial class OAVariantLib
         #region Constants
 
         // Constants for VariantChangeType from OleAuto.h
-        public const int NoValueProp = 0x01;
-        public const int AlphaBool = 0x02;
-        public const int NoUserOverride = 0x04;
-        public const int CalendarHijri = 0x08;
         public const int LocalBool = 0x10;
 
-        internal static readonly Type?[] ClassTypes = {
-            typeof(Empty),
-            typeof(void),
-            typeof(bool),
-            typeof(char),
-            typeof(sbyte),
-            typeof(byte),
-            typeof(short),
-            typeof(ushort),
-            typeof(int),
-            typeof(uint),
-            typeof(long),
-            typeof(ulong),
-            typeof(float),
-            typeof(double),
-            typeof(string),
-            typeof(void),
-            typeof(DateTime),
-            typeof(TimeSpan),
-            typeof(object),
-            typeof(decimal),
-            null,  // Enums - what do we do here?
-            typeof(Missing),
-            typeof(DBNull),
+        private static readonly Dictionary<Type, VarEnum> ClassTypes = new Dictionary<Type, VarEnum>
+        {
+            { typeof(Empty), VarEnum.VT_EMPTY },
+            { typeof(void), VarEnum.VT_VOID },
+            { typeof(bool), VarEnum.VT_BOOL },
+            { typeof(char), VarEnum.VT_I2 },
+            { typeof(sbyte), VarEnum.VT_I1 },
+            { typeof(byte), VarEnum.VT_UI1 },
+            { typeof(short), VarEnum.VT_I2 },
+            { typeof(ushort), VarEnum.VT_UI2 },
+            { typeof(int), VarEnum.VT_I4 },
+            { typeof(uint), VarEnum.VT_UI4 },
+            { typeof(long), VarEnum.VT_I8 },
+            { typeof(ulong), VarEnum.VT_UI8 },
+            { typeof(float), VarEnum.VT_R4 },
+            { typeof(double), VarEnum.VT_R8 },
+            { typeof(string), VarEnum.VT_BSTR },
+            { typeof(DateTime), VarEnum.VT_DATE },
+            { typeof(object), VarEnum.VT_UNKNOWN },
+            { typeof(decimal), VarEnum.VT_DECIMAL },
```

Comment:
Just a quick check, did the newly added tests have a case for each of these values?

Reply:
Not quite. This list is taken as-is from the old table, modulo types rejected at native side.
The types need to be present in signature. I'm a bit unsure about what would happen to `Empty`, or `void` return.
To be exhaustive, it would need every type existing at source and target. Is it necessary to cover every primitive number type? Or just the "special" interesting types?

Reply:
I'm not concerned with the primitives - i1-i8, u1-u8, r4, r8 - or the other built-in types - `BSTR`, `IUnknown` - the other OLE based ones always give me pause, because rarely do they do what I would expect. I want to ensure that we have basic coverage of the semantics and then we can enrich the cases as needed or make informed decisions if we decide to change/duplicate their behavior going forward.

Reply:
I've updated the tests. They should now be exhaustive about all non-primitive types that's passed/accepted here. I ran the tests on main branch.
Also removed some more impossible types. `void` won't reach this path. `Empty` is an internal type and won't appear in public signature. `object` will always pass the `IsInstanceOfType` check and doesn't need conversion.